compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala

```diff
@@ -99,10 +99,17 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
      * because they are emitted as private by our .scala source files, but
      * they are considered public at use site since their symbols come from
      * Java-emitted .class files.
+     * We also special case outer fields because with multiple inheritance it
+     * is possible to have several definitions of such fields in the same
+     * instance, one for each class/trait that captures its outer context.
+     * The name for the outer field is always `$outer` which could potentially
+     * collapse into a single field.
      */
-    val idSuffix =
+    val idSuffix: String =
       if (sym.isPrivate || allRefClasses.contains(sym.owner))
         sym.owner.ancestors.count(!_.isTraitOrInterface).toString
+      else if (sym.isOuterField)
+        sym.owner.name.toString
```

Comment:
You should use `encodeClassFullName(sym.owner)` so that weird characters in `sym.owner.name` are dealt with.